src/mqtt_client/mqtt_client.rs

```diff
@@ -83,18 +107,81 @@ where
             )
             .unwrap();
 
-        let mut settings_prefix: String<64> = String::from(prefix);
+        let mut settings_prefix: String<MAX_PREFIX_LENGTH> = String::from(prefix);
         settings_prefix.push_str("/settings").unwrap();
 
         Ok(Self {
             mqtt,
-            settings: Settings::default(),
+            clock,
+            settings,
             settings_prefix,
             prefix: String::from(prefix),
             subscribed: false,
+            iteration_state: None,
+            rx_timeout: None,
         })
     }
 
+    fn handle_republish(&mut self) {
+        if let Some(timeout) = &self.rx_timeout {
+            if self.clock.try_now().unwrap() > *timeout {
+                log::info!("Republish duration lapsed. Publishing settings");
+                self.iteration_state.replace([0; 16]);
+            }
+        }
+
+        let mut iteration_exhausted = false;
+        if let Some(ref mut iteration_state) = &mut self.iteration_state {
+            self.rx_timeout.take();
+
+            if !self.mqtt.client.can_publish(QoS::AtMostOnce) {
+                return;
+            }
+
+            for topic in self
+                .settings
+                .into_iter::<{ MAX_TOPIC_LENGTH }>(iteration_state)
+                .unwrap()
+            {
+                let mut data = [0; MESSAGE_SIZE];
+
+                // Note(unwrap): We know this topic exists already because we just got it from the
+                // iterator.
+                let len = self.settings.get(&topic, &mut data).unwrap();
+
+                let mut prefixed_topic: String<{ MAX_TOPIC_LENGTH }> = String::new();
+                write!(&mut prefixed_topic, "{}/{}", &self.settings_prefix, &topic).unwrap();
+
+                // Note(unwrap): This should not fail because `can_publish()` was checked before
+                // attempting this publish.
+                self.mqtt
+                    .client
+                    .publish(
+                        &prefixed_topic,
+                        &data[..len],
+                        QoS::AtMostOnce,
+                        Retain::NotRetained,
+                        &[],
+                    )
+                    .unwrap();
+
+                // If we can't publish any more messages, bail out now to prevent the iterator from
+                // progressing. If we don't bail out now, we'd silently drop a setting.
+                if !self.mqtt.client.can_publish(QoS::AtMostOnce) {
+                    return;
+                }
+            }
+
+            iteration_exhausted = true;
+        }
+
+        // If iteration has been exhausted, clear the iterator state, as we are now done
+        // republishing.
+        if iteration_exhausted {
+            self.iteration_state.take();
+        }
+    }
```

Comment:
I don't think you need `iteration_exhausted`. The only paths out of the loop is either return or completing. The latter then must `take()` in all cases.

And shouldn't we more cleanly store an `Option<Iterator> and not its opaque internal state? I would create and own the iterator state in the iterator struct.

Reply:
The iterator struct cannot exist outside of this function because it has to borrow a reference to `settings`. However, `settings` is mutably borrowed in `update()`, so we need to release any immutable references to it inbetween. That's why the iterator is intended to go out of scope when this function exits, but the iteration information is retained in `state` so we can resume from where we left off.

Reply:
Ah. Interesting problem. I would still make the state more opaque though and ensure type coherence between saving and resuming. Then maybe more generically a trait `ResumableIterator<T>` with `save(self) -> Self::State` and `resume(Self::State) -> Self`?

Reply:
I would personally prefer if we pulled discussion of that out of this PR and into a refactor of the `Miniconf` trait if we feel it to be necessary. Right now, the iterator state is not exposed to the end user, so there's no need to typestate it. It will always be assocaited with the `MqttClient::Settings` type and is internal only.

Reply:
Ok. But I still have the suspicion that the following are common problems where it is somewhat likely that someone else has solved them: (a) suspending and resuming an iterator (b) iterating over the fields of a struct (there seems to be [`field_names`](https://github.com/TedDriggs/field_names)). Especially all this introspection stuff sounds related.
I'd rather evaluate those other solutions early than later having/wanting to migrate.

Reply:
Stringifying the names is trivial in a proc-macro. That is _not_ what we have done, and is not where the complexity of this solution comes from.

What we have done is allowed recursive iteration of nested structures by mapping unique indices to each field (in a recursive manner). 

James and I evaluated a number of various crates and found that none did _quite_ what we wanted. Similarly, this isn't necessarily an issue of just pausing/resuming iteration that can be solved easily because this has to fight with Rust's ownership system on the thing we're iterating. AFAICT there is no existing `SuspendableIterator` trait.

While we could make one, I don't think it actually either helps us significantly

Reply:
If you want to look at the change set related to _just_ iteration over the settings structure, the PR is https://github.com/quartiq/miniconf/pull/62

Reply:
Looking into `field_names` looks like it'd be somewhat helpful, but it'd also require us to rely on the user using `#[derive(FieldNames, Miniconf)]` because we cannot invoke proc macros from other proc macros. If they forgot the derive, It's likely they'd get extremely opaque error messages about not being able to find T::FIELDS

Reply:
Ack. I haven't seen the details of evaluation and I can't find it in the PR.
On a quick search I also was unable to find something like a `SuspendableIterator` trait but maybe you agree that this is potentially a worthwhile generalization.
I do think generalizing these would help even if it only forces the analysis of the consistency and soundness of this design.
For now as an MVP this is ok. But my feeling is that it will either break or migrate in the mid-term.

Reply:
I've opened an issue into investigating already-created suspendable iteration as https://github.com/quartiq/miniconf/issues/69